### PR TITLE
Feat: estimate bundle gas

### PR DESCRIPTION
**Pull request by datluongductuan**

### CODE REVIEW

File: pkg/mev/bundle_sender.go [hunk 1]
```diff
@@ -124,6 +134,35 @@ func (s *Client) flashbotBackrunSendBundle(
...
+	}
+
+	var gasEstimateCost []hexutil.Uint64
+
+	err := s.ethClient.Client().Call(
+		&gasEstimateCost, ETHEstimateGasBundleMethod,
+		map[string]interface{}{
+			"transactions": bundles,
+		}, "latest", overrides,
+	)
```
Comment (hiepnv90) [inline pkg/mev/bundle_sender.go:154]: Estimate gas for bundle is not supported by node, why need to call to `ethClient` to estimate gas for bundle? Why don't use a builder client to estimate it?

Comment (datluongductuan) [reply pkg/mev/bundle_sender.go:154]: Using estimate gas for the bundle by builder client requires signed txs (using `eth_callBundle`). But the current flow requires we estimate gas before signing Tx (https://github.com/KyberNetwork/poseidon/blob/23075396687934c69eaa21507577255170220562/internal/executor/blockchain/blockchain.go#L959)

So if I estimate gas for backrun tx, I might got return amount is not enough. And then I can't have a signed tx to estimate gas bundle.

So I decided to use our trading node's custom estimate gas bundle rpc to estimate gas bundle without signed txs (only need `ethereum.CallMsg`).

Pls correct me if have another way

Comment (hiepnv90) [reply pkg/mev/bundle_sender.go:154]: So you should define an interface for estimating a bundle with overrides. I'm also thinking about a way to use `eth_estimateGas` with overrides pools' state for estimating gas for backrun tx.

Comment (datluongductuan) [reply pkg/mev/bundle_sender.go:154]:
```
type IBundleSender interface {
	.....
	EstimateBundleGas(
		ctx context.Context,
		messages []ethereum.CallMsg,
		overrides *map[common.Address]gethclient.OverrideAccount,
	) ([]uint64, error)
	....
}
```

I have defined the `EstimateBundleGas` func signature in the IBundleSender interface. You mean I should define another new interface for estimating a bundle right?

Comment (datluongductuan) [reply pkg/mev/bundle_sender.go]: Thanks. I've defined a new interface and refactor by following this way

File: pkg/mev/gas_bundle_estimator.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/ethclient/gethclient"
+)
+
+type GasBundleEstimator struct {
+	ethClient *ethclient.Client
```
Comment (hiepnv90) [inline pkg/mev/gas_bundle_estimator.go:14]: Should be `rpc.Client` instead of `ethclient.Client`

Comment (datluongductuan) [reply pkg/mev/gas_bundle_estimator.go]: fixed

**Review by hiepnv90 — APPROVED** [review]

**Pull request merged**